pkg/chain/bitcoin/electrs.go

```diff
@@ -38,54 +47,98 @@ func (e *ElectrsConnection) setClient(client httpClient) {
+		if err != nil {
+			return fmt.Errorf("something went wrong reading the electrs response body: [%v]", err)
+		}
+		if bytesCopied == 0 {
+			return fmt.Errorf("something went wrong reading the electrs response body: 0 bytes copied")
+		}
+		logger.Infof("successfully broadcast the bitcoin transaction: %s", transactionIDBuffer.String())
 		return nil
-	}
-	resp, err := e.client.Post(fmt.Sprintf("%s/tx", e.apiURL), "text/plain", strings.NewReader(transaction))
-	if err != nil {
-		return err
-	}
-	if resp.StatusCode != 200 {
-		return fmt.Errorf("something went wrong with broadcast: [%s], transaction: [%s]", resp.Status, transaction)
-	}
-	transactionIDBuffer := new(strings.Builder)
-	bytesCopied, err := io.Copy(transactionIDBuffer, resp.Body)
-	// if the status code was 200, but we were unable to read the body, log an
-	// error but return successfully anyway.
-	if err != nil {
-		logger.Errorf("something went wrong reading the electrs response body: [%v]", err)
-	}
-	if bytesCopied == 0 {
-		logger.Error("something went wrong reading the electrs response body: 0 bytes copied")
-	}
-	logger.Infof("successfully broadcast the bitcoin transaction: %s", transactionIDBuffer.String())
-	return nil
+	})
 }
 
-// VbyteFee retrieves the 25-block estimate fee per vbyte on the bitcoin network.
-func (e ElectrsConnection) VbyteFee() (int32, error) {
+// VbyteFeeFor25Blocks retrieves the 25-block estimate fee per vbyte on the bitcoin network.
+func (e ElectrsConnection) VbyteFeeFor25Blocks() (int32, error) {
 	if e.apiURL == "" {
-		return 0, nil
+		return 0, fmt.Errorf("attempted to call VbyteFee with no apiURL")
 	}
-	resp, err := e.client.Get(fmt.Sprintf("%s/fee-estimates", e.apiURL))
+	var vbyteFee int32
+	err := utils.DoWithDefaultRetry(e.timeout, func(ctx context.Context) error {
+		resp, err := e.client.Get(fmt.Sprintf("%s/fee-estimates", e.apiURL))
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode != 200 {
+			return fmt.Errorf("something went wrong with broadcast: [%s]", resp.Status)
+		}
+
+		var fees map[string]float32
+		err = json.NewDecoder(resp.Body).Decode(&fees)
+		if err != nil {
+			return fmt.Errorf("something went wrong decoding the vbyte fees: [%v]", err)
+		}
+		fee, ok := fees["25"]
+		if !ok {
+			fee = 0
+		}
+		logger.Info("retrieved a vbyte fee of [%d]", fee)
+		vbyteFee = int32(fee)
+		return nil
+	})
 	if err != nil {
 		return 0, err
 	}
-	if resp.StatusCode != 200 {
-		return 0, fmt.Errorf("something went wrong with broadcast: [%s]", resp.Status)
+	return vbyteFee, nil
+}
+
+func (e ElectrsConnection) IsAddressUnused(btcAddress string) (bool, error) {
+	if e.apiURL == "" {
+		return false, fmt.Errorf("attempted to call IsAddressUnused with no apiURL")
 	}
+	isAddressUnused := false
+	err := utils.DoWithDefaultRetry(e.timeout, func(ctx context.Context) error {
+		resp, err := e.client.Get(fmt.Sprintf("%s/address/%s/txs", e.apiURL, btcAddress))
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode != 200 {
```

Comment:
In `Broadcast` and `VbyteFeeFor25Blocks` functions we return an error right after checking `resp.StatusCode != 200`. Here we're trying to resolve the error message. Is there a reason we have two different approaches? Could we align them all?

Reply:
For broadcast and vbyte fee, the errors are shipped back in the status, so all we need to read is the status. `IsAddressUnused` has a more complicated API because there's more cases it has to handle (like sending an invalid btc address in the url param). The API ships those errors back in the payload along with a non-200 status code.

If you try to hit `localhost:3002/address/banana/tx`, for example, you get a 400 response with a payload of "Invalid Bitcoin address", and that error doesn't show up in the status.